Description: Build a shared library properly.
 - build the library as libsdb.so.0.0.0
 - pass the -soname parameter to the linker with version 0
 - break out the compiler and linker flags for the shared build
 - compile separate object files for the shared library
Forwarded: not yet
Author: Peter Pentchev <roam@ringlet.net>
Last-Update: 2013-03-27

--- a/config.mk
+++ b/config.mk
@@ -8,6 +8,7 @@
 CFLAGS+=-Wall
 #CFLAGS+=-O3
 #CFLAGS+=-ggdb -g -Wall -O0
+CFLAGS_SHARED?=-fPIC -shared
 
 HAVE_VALA=$(shell valac --version)
 # This is hacky
@@ -20,7 +21,10 @@
    CC+=-arch i386 -arch x86_64
  endif
 else
-SOEXT=so
+SOVERSION=0
+SOEXT=so.0.0.0
+LDFLAGS_SHARED?=-fPIC -shared
+LDFLAGS_SHARED+=-Wl,-soname,libsdb.so.$(SOVERSION)
 endif
 RANLIB?=ranlib
 EXEXT=
--- a/src/Makefile
+++ b/src/Makefile
@@ -5,6 +5,8 @@
 OBJ=cdb.o buffer.o cdb_make.o ls.o ht.o sdb.o sdbn.o sdba.o query.o
 OBJ+=json.o json/js0n.o json/json.o json/rangstr.o ns.o lock.o
 
+SOBJ=$(subst .o,.so,${OBJ})
+
 BIN=sdb${EXEXT}
 
 .PHONY: all static shared clean
@@ -28,8 +30,11 @@
 	${AR} -r libsdb.a ${OBJ}
 	${RANLIB} libsdb.a
 
-libsdb.${SOEXT}: ${OBJ}
-	${CC} ${CFLAGS} ${LDFLAGS} -o $@ $(subst .o,.c,${OBJ}) -fPIC -shared
+%.so:	%.c
+	${CC} -c ${CPPFLAGS} ${CFLAGS} ${CFLAGS_SHARED} -o $@ $^
+
+libsdb.${SOEXT}: ${SOBJ}
+	${CC} ${LDFLAGS} $(LDFLAGS_SHARED) -o $@ ${SOBJ}
 
 t:
 	rm -f foo
